### PR TITLE
[Merged by Bors] - feat: port Init.Data.Subtype.Basic

### PR DESCRIPTION
**Pull request by j-loreaux**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Init/Data/Subtype/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+#align subtype.tag_irrelevant Subtype.tag_irrelevant
+
+theorem ne_of_val_ne {a1 a2 : { x // p x }} : val a1 ≠ val a2 → a1 ≠ a2 :=
+  mt <| congr_arg _
+#align subtype.ne_of_val_ne Subtype.ne_of_val_ne
+
+end Subtype
+
+open Subtype
+
```
Comment (j-loreaux) [inline Mathlib/Init/Data/Subtype/Basic.lean:41]:
```suggestion

/-- If there is some element satisfying the predicate, then the subtype is inhabited. -/
```

File: Mathlib/Init/Data/Subtype/Basic.lean [hunk 2]
```diff
@@ -27,27 +31,15 @@ theorem tag_irrelevant {a : α} (h1 h2 : p a) : mk a h1 = mk a h2 :=
   rfl
 #align subtype.tag_irrelevant Subtype.tag_irrelevant
 
-#print Subtype.eq /-
-protected theorem eq : ∀ {a1 a2 : { x // p x }}, val a1 = val a2 → a1 = a2
-  | ⟨x, h1⟩, ⟨x, h2⟩, rfl => rfl
-#align subtype.eq Subtype.eq
```
Comment (jcommelin) [inline Mathlib/Init/Data/Subtype/Basic.lean]: Please keep the align statement.

File: Mathlib/Init/Data/Subtype/Basic.lean [hunk 3]
```diff
@@ -27,27 +31,15 @@ theorem tag_irrelevant {a : α} (h1 h2 : p a) : mk a h1 = mk a h2 :=
...
--/
-
 theorem ne_of_val_ne {a1 a2 : { x // p x }} : val a1 ≠ val a2 → a1 ≠ a2 :=
   mt <| congr_arg _
 #align subtype.ne_of_val_ne Subtype.ne_of_val_ne
 
-#print Subtype.eta /-
-theorem eta (a : { x // p x }) (h : p (val a)) : mk (val a) h = a :=
-  Subtype.eq rfl
-#align subtype.eta Subtype.eta
```
Comment (jcommelin) [inline Mathlib/Init/Data/Subtype/Basic.lean]: Idem.

File: Mathlib/Init/Data/Subtype/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+theorem exists_of_subtype {α : Type u} {p : α → Prop} : { x // p x } → ∃ x, p x
+  | ⟨a, h⟩ => ⟨a, h⟩
+#align subtype.exists_of_subtype Subtype.exists_of_subtype
+
+variable {α : Type u} {p : α → Prop}
+
+theorem tag_irrelevant {a : α} (h1 h2 : p a) : mk a h1 = mk a h2 :=
+  rfl
+#align subtype.tag_irrelevant Subtype.tag_irrelevant
+
```
Comment (j-loreaux) [inline Mathlib/Init/Data/Subtype/Basic.lean:33]:
```suggestion

#align subtype.eq Subtype.eq
```

File: Mathlib/Init/Data/Subtype/Basic.lean [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+theorem ne_of_val_ne {a1 a2 : { x // p x }} : val a1 ≠ val a2 → a1 ≠ a2 :=
+  mt <| congr_arg _
+#align subtype.ne_of_val_ne Subtype.ne_of_val_ne
+
```
Comment (j-loreaux) [inline Mathlib/Init/Data/Subtype/Basic.lean:37]:
```suggestion

#align subtype.eta Subtype.eta

```

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**